packages/cache/src/lib/CacheMiddleware.ts

```diff
@@ -0,0 +1,727 @@
+import {
+  ApplyMiddlewareOptions,
+  isEveryByHookPayload,
+  isEveryByValuePayload,
+  isFilterByHookPayload,
+  isFilterByValuePayload,
+  isFindByHookPayload,
+  isFindByValuePayload,
+  isMapByHookPayload,
+  isMapByPathPayload,
+  isPayloadWithData,
+  isRemoveByHookPayload,
+  isRemoveByValuePayload,
+  JoshMiddleware,
+  JoshMiddlewareStore,
+  JoshProvider,
+  Method,
+  Payload,
+  Payloads,
+  PostProvider,
+  PreProvider
+} from '@joshdb/provider';
+import { addExitCallback } from 'catch-exit';
+import { getProperty } from 'property-helpers';
+
+@ApplyMiddlewareOptions({ name: 'cache' })
+export class CacheMiddleware<StoredValue = unknown> extends JoshMiddleware<CacheMiddleware.ContextData<StoredValue>, StoredValue> {
+  public pollingInterval: NodeJS.Timer | undefined;
+
+  public async init(store: JoshMiddlewareStore<StoredValue>) {
+    await super.init(store);
+    await this.fetchCache();
+
+    return this;
+  }
+
+  @PreProvider()
+  public async [Method.Get]<Value = StoredValue>(payload: Payloads.Get<Value>): Promise<Payloads.Get<Value>> {
+    const { key, path } = payload;
+    const { provider: cache } = this.context;
+    const getPayload = await cache[Method.Get]<CacheMiddleware.Document<Value>>({ method: Method.Get, key, path, errors: [] });
+
+    payload.errors = [...payload.errors, ...getPayload.errors];
+
+    if (isPayloadWithData<CacheMiddleware.Document<Value>>(getPayload)) {
+      const { data } = getPayload;
+
+      if (await this.checkNotExpired(data, key)) payload.data = data.value;
+    }
+
+    return payload;
+  }
+
+  @PreProvider()
+  public async [Method.Each](payload: Payloads.Each<StoredValue>): Promise<Payloads.Each<StoredValue>> {
+    const { hook } = payload;
+    const { provider: cache } = this.context;
+    const cacheHook = async (value: CacheMiddleware.Document<StoredValue>, key: string) => {
+      if (await this.checkNotExpired(value, key)) {
+        return hook(value.value, key);
+      }
+
+      const bypassPayload = await this.provider[Method.Get]({ method: Method.Get, key, path: [], errors: [] });
+
+      if (isPayloadWithData<StoredValue>(bypassPayload)) {
+        const { data } = bypassPayload;
+
+        return hook(data, key);
+      }
+    };
+
+    const { errors } = await cache[Method.Each]({ method: Method.Each, hook: cacheHook, errors: [] });
```

Comment:
Why is this being called if there is no data being collected from it?

Reply:
Since it's .Each, it's running the hook on each value, so it doesn't return data

Reply:
Oh wait, thought this was a different method 😄